File: backend/app/api/community.py
"""Community Knowledge public endpoints."""

from __future__ import annotations

from typing import Any
from fastapi import APIRouter, Header, HTTPException, Query, Request, status
from pydantic import BaseModel, Field

from app.repositories.community import CommunityRepository, DuplicatePerspectiveError
from app.services.community_service import CommunityService

router = APIRouter(prefix="/api/v2/community", tags=["community"])


class ContributionCreateRequest(BaseModel):
    cue: str = Field(..., min_length=1, max_length=100)
    scenario: str = Field(..., min_length=5, max_length=1000)
    scenario_family: str = Field(..., min_length=1, max_length=50)
    relationship: str = Field(..., min_length=1, max_length=50)
    channel: str = Field(..., min_length=1, max_length=50)
    formality: str = Field(..., min_length=1, max_length=50)
    interpretation: str = Field(..., min_length=5, max_length=1000)
    conditions: str = Field("", max_length=500)
    counterconditions: str = Field("", max_length=500)
    do_not_assume: str = Field("", max_length=500)
    safe_action: str = Field("", max_length=500)
    counterexample: str = Field("", max_length=500)
    evidence_scope: str = Field("", max_length=500)
    direct_experience: bool = True
    consent_given: bool = True


class PerspectiveCreateRequest(BaseModel):
    interpretation: str = Field(..., min_length=5, max_length=1000)
    conditions: str = Field("", max_length=500)
    counterconditions: str = Field("", max_length=500)


def get_community_service(request: Request) -> CommunityService:
    repo = CommunityRepository(request.app.state.database)
    return CommunityService(repo)


def get_community_repo(request: Request) -> CommunityRepository:
    return CommunityRepository(request.app.state.database)


@router.post("/contributions", status_code=status.HTTP_201_CREATED)
async def submit_contribution(
    req: ContributionCreateRequest,
    request: Request,
    x_contributor_id: str = Header(..., alias="X-Contributor-ID"),
    idempotency_key: str | None = Header(None, alias="Idempotency-Key"),
):
    if not x_contributor_id or not x_contributor_id.strip():
        raise HTTPException(status_code=400, detail="X-Contributor-ID header is required")

    service = get_community_service(request)
    result = service.submit_contribution(
        x_contributor_id, req.model_dump(), idempotency_key=idempotency_key
    )
    if not result.get("success"):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"error": result.get("error"), "privacy_flags": result.get("flags")},
        )
    return result["data"]


@router.get("/candidates")
async def list_candidates(
    request: Request,
    status: str | None = Query(None),
    limit: int = Query(20, ge=1, le=100),
    cursor: int = Query(0, ge=0),
):
    repo = get_community_repo(request)
    return repo.list_public_candidates(status=status, limit=limit, cursor=cursor)


@router.get("/candidates/similar")
async def find_similar_candidates(
    request: Request,
    cue: str = Query(..., min_length=1),
    scenario_family: str = Query("", max_length=50),
    limit: int = Query(5, ge=1, le=20),
):
    repo = get_community_repo(request)
    return repo.find_similar_candidates(cue=cue, scenario_family=scenario_family, limit=limit)


@router.get("/candidates/{candidate_id}")
async def get_candidate(candidate_id: str, request: Request):
    repo = get_community_repo(request)
    cand = repo.get_public_candidate(candidate_id)
    if not cand:
        raise HTTPException(status_code=404, detail="Candidate not found")
    return cand


@router.post("/candidates/{candidate_id}/perspectives", status_code=status.HTTP_201_CREATED)
async def add_perspective(
    candidate_id: str,
    req: PerspectiveCreateRequest,
    request: Request,
    x_contributor_id: str = Header(..., alias="X-Contributor-ID"),
):
    if not x_contributor_id or not x_contributor_id.strip():
        raise HTTPException(status_code=400, detail="X-Contributor-ID header is required")

    service = get_community_service(request)
    try:
        result = service.add_perspective(candidate_id, x_contributor_id, req.model_dump())
    except DuplicatePerspectiveError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))

    if not result.get("success"):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"error": result.get("error"), "privacy_flags": result.get("flags")},
        )
    return result["data"]


@router.get("/contributions/{contribution_id}/status")
async def get_contribution_status(contribution_id: str, request: Request):
    repo = get_community_repo(request)
    receipt = repo.get_status_receipt(contribution_id)
    if not receipt:
        raise HTTPException(status_code=404, detail="Contribution receipt not found")
    return receipt

File: frontend/src/services/api/community.ts
import { getOrCreateContributorId } from "../storage/contributor";

export interface ContributionPayload {
  cue: string;
  scenario: string;
  scenario_family: string;
  relationship: string;
  channel: string;
  formality: string;
  interpretation: string;
  conditions?: string;
  counterconditions?: string;
  do_not_assume?: string;
  safe_action?: string;
  counterexample?: string;
  evidence_scope?: string;
  direct_experience?: boolean;
  consent_given?: boolean;
}

export interface PerspectivePayload {
  interpretation: string;
  conditions?: string;
  counterconditions?: string;
}

export interface CandidatePublic {
  candidate_id: string;
  cue: string;
  scenario: string;
  scenario_family: string;
  relationship: string;
  channel: string;
  formality: string;
  status: string;
  provenance: string;
  do_not_assume: string;
  safe_action: string;
  counterexample: string;
  evidence_scope: string;
  perspectives: Array<{
    perspective_index: number;
    interpretation: string;
    conditions: string;
    counterconditions: string;
  }>;
  readiness: {
    ready: boolean;
    missing_requirements: string[];
    perspective_count: number;
    distinct_contributors: number;
  };
}

export interface ContributionReceipt {
  contribution_id: string;
  candidate_id: string;
  status: string;
  provenance: string;
  readiness: {
    ready: boolean;
    missing_requirements: string[];
    perspective_count: number;
    distinct_contributors: number;
  };
  created_at: string;
}

export async function submitContribution(payload: ContributionPayload): Promise<ContributionReceipt> {
  const contributorId = getOrCreateContributorId();
  const res = await fetch("/api/v2/community/contributions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Contributor-ID": contributorId,
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const errorData = await res.json().catch(() => ({}));
    throw new Error(errorData.detail?.error || errorData.detail || `Server error: ${res.status}`);
  }
  return res.json();
}

export async function listCandidates(status?: string): Promise<CandidatePublic[]> {
  const url = status ? `/api/v2/community/candidates?status=${encodeURIComponent(status)}` : "/api/v2/community/candidates";
  const res = await fetch(url);
  if (!res.ok) throw new Error("Failed to fetch candidates");
  return res.json();
}

export async function findSimilarCandidates(cue: string, scenario_family = ""): Promise<CandidatePublic[]> {
  const params = new URLSearchParams({ cue, scenario_family });
  const res = await fetch(`/api/v2/community/candidates/similar?${params.toString()}`);
  if (!res.ok) return [];
  return res.json();
}

export async function getCandidate(candidateId: string): Promise<CandidatePublic> {
  const res = await fetch(`/api/v2/community/candidates/${encodeURIComponent(candidateId)}`);
  if (!res.ok) throw new Error("Candidate not found");
  return res.json();
}

export async function addPerspective(candidateId: string, payload: PerspectivePayload): Promise<any> {
  const contributorId = getOrCreateContributorId();
  const res = await fetch(`/api/v2/community/candidates/${encodeURIComponent(candidateId)}/perspectives`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Contributor-ID": contributorId,
    },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to add perspective");
  }
  return res.json();
}

export async function getContributionStatus(contributionId: string): Promise<any> {
  const res = await fetch(`/api/v2/community/contributions/${encodeURIComponent(contributionId)}/status`);
  if (!res.ok) throw new Error("Receipt not found");
  return res.json();
}
